genie-web/src/test/groovy/com/netflix/genie/web/services/impl/JobDirectoryServerServiceImplSpec.groovy

```diff
@@ -0,0 +1,150 @@
+
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+class JobDirectoryServerServiceImplSpec extends Specification {
+    static final String JOB_ID = "123456"
+    static final String relPath = "bar/foo.txt"
+    static final URL BASE_URL = new URL("https", "genie.netflix.net", 8080, "/jobs/" + JOB_ID + "/output/" + relPath)
+    static final URI EXPECTED_V4_FILE_URI = new URI(AgentFileProtocolResolver.URI_SCHEME, JOB_ID, "/" + relPath, null);
+
+    ResourceLoader resourceLoader
+    JobPersistenceService jobPersistenceService
+    JobFileService jobFileService
+    AgentFileStreamService agentFileStreamService
+    MeterRegistry meterRegistry
+    JobDirectoryServerService service
+    HttpServletRequest request
+    HttpServletResponse response
+    JobDirectoryManifest manifest
+    JobDirectoryManifest.ManifestEntry manifestEntry
+    Resource resource
+    JobDirectoryServerServiceImpl.GenieResourceHandler.Factory handlerFactory
+    JobDirectoryServerServiceImpl.GenieResourceHandler handler
+
+    void setup() {
+        this.resourceLoader = Mock(ResourceLoader)
+        this.jobPersistenceService = Mock(JobPersistenceService)
+        this.jobFileService = Mock(JobFileService)
+        this.agentFileStreamService = Mock(AgentFileStreamService)
+        this.meterRegistry = Mock(MeterRegistry)
+        this.handlerFactory = Mock(JobDirectoryServerServiceImpl.GenieResourceHandler.Factory)
+        this.handler = Mock(JobDirectoryServerServiceImpl.GenieResourceHandler)
+        this.service = new JobDirectoryServerServiceImpl(resourceLoader, jobPersistenceService, jobFileService, agentFileStreamService, meterRegistry, handlerFactory)
+
+        this.request = Mock(HttpServletRequest)
+        this.response = Mock(HttpServletResponse)
+        this.manifest = Mock(JobDirectoryManifest)
+        this.manifestEntry = Mock(JobDirectoryManifest.ManifestEntry)
+        this.resource = Mock(Resource)
+    }
+
+    def "ServeResource -- job not found (status)"() {
+        setup:
+        Exception e = new GenieNotFoundException("...")
+
+        when:
+        service.serveResource(JOB_ID, BASE_URL, relPath, request, response)
+
+        then:
+        1 * jobPersistenceService.getJobStatus(JOB_ID) >> { throw e }
+        1 * response.sendError(404, e.getMessage())
+    }
+
+    def "ServeResource -- job not found (v4)"() {
+        setup:
+        Exception e = new GenieJobNotFoundException("...")
+
+        when:
+        service.serveResource(JOB_ID, BASE_URL, relPath, request, response)
+
+        then:
+        1 * jobPersistenceService.getJobStatus(JOB_ID) >> JobStatus.RUNNING
+        1 * jobPersistenceService.isV4(JOB_ID) >> { throw e }
+        1 * response.sendError(404, e.getMessage())
+    }
+
+    def "ServeResource -- invalid URI"() {
+        // TODO -- how to cause URI syntax exception?
+    }
+
+    def "ServeResource -- Active V4 job, manifest not found"() {
+        setup:
```

Comment:
FYI setup isn't really necessary. Anything before the first `when` I believe is considered setup